### PR TITLE
refactor kafka sink into a state machine

### PR DESCRIPTION
**Pull request by elindsey**
Motivation is to lay the groundwork for transactional writing, plus simplify the code/remove a few levels of nesting. 

Transactional writing is significantly easier to implement as a state machine, adding phases for init/begin txn/commit txn/abort - if you want to see the direction, the current prototype of txn writes that builds on this diff is here https://github.com/elindsey/materialize/blob/once/src/dataflow/src/sink/kafka.rs#L415

This introduces one major functional change to the non-txn path: we don't write results for a given timestamp until that timestamp has been closed. It also introduces one minor change: the consistency topic is no longer interleaved as we service a single timestamp at a time.

This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5166)

### CODE REVIEW

File: src/dataflow/src/sink/kafka.rs [hunk 1]
```diff
@@ -237,283 +219,243 @@ where
...
+                    key,
+                    value,
+                    count: diff.abs() as usize,
                 });
+                s.metrics.rows_queued.inc();
+            }
+        });
+
+        // Move any newly closed timestamps from pending to ready
+        let mut closed_ts: Vec<u64> = pending_rows
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:330]: I feel like BTreeMap's `drain_filter` does exactly what you want here but its nightly only :(

**Review by ruchirK — APPROVED** [review]
Nice! This seems like a pretty huge improvement over the existing sink implementation

File: src/dataflow/src/sink/kafka.rs (hunk 1 above)
Comment (elindsey) [reply src/dataflow/src/sink/kafka.rs:330]: It does; I look forward to replacing this whole bit eventually :/

Comment (quodlibetor) [reply src/dataflow/src/sink/kafka.rs:330]: We could just copy the `drain_filter` code from std into our `ore`, it seems like less than a few dozen lines of code. Which is more than you've got here, but might allow for reuse/weighing in on the stabilization issue? Not sure if we would want it anywhere else, though.

File: src/dataflow/src/sink/kafka.rs [hunk 2]
```diff
@@ -125,74 +120,67 @@ impl ProducerContext for SinkProducerContext {
...
-            .map(|(&k, &v)| (k, v))
-            .collect();
-
-        for (time, _) in closed_timestamps.iter() {
-            self.timestamp_counts.remove(time);
-        }
+enum SendState {
+    Begin,
+    Draining {
+        row_index: usize,
```
[the remark this answers is not in the document]
Comment (elindsey) [reply src/dataflow/src/sink/kafka.rs:176]: Good point! I added a few comments. I'll flesh this out even more in the transactional diff - that'll complicate things a bit

File: src/dataflow/src/sink/kafka.rs (hunk 1 above)
Comment (elindsey) [reply src/dataflow/src/sink/kafka.rs:337]: It relies a little too much on the internals of btreemap than I'm comfortable with - I don't really want to deal with keeping it in sync with upstream even if all the APIs it uses are public.

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:337]: yeah I agree with Eli here.

**Comment by elindsey** [comment]
Doing a quick rebase/test run prior to landing

**Pull request merged**